man/nodelist.1
.TH "nodelist" 1
.SH "NAME"

nodelist

.SH "SYNOPSIS"
.B nodelist [options] 

.SH "DESCRIPTION"
.TP
List resources on the cluster system.

.SH "OPTIONS"
.TP
.B \-b <list of node ids>
.B [Cray Only]
Print detailed information for all nodes in the list of node ids.
This will accept hyphenated ranges as well.  All ranges are inclusive.
.TP
.B \-d \-\-debug
Turn on communication debugging.
.TP
.B \-h \-\-help
Displays the usage and a brief descriptions of the options
.TP
.B \-\-version
Displays client revision and Cobalt version

.SH "SEE ALSO"
.BR nodeadm (8)
.SH "BUGS"


CHANGES
= Changes from previous Cobalt Versions =

== Changes to 1.0.6 ==
 * Backfill now supported on Cray systems.
 * Major performance improvement to nodelist and nodeadm -l.
 * Fix for an error that can cause loss of job tracking during job startup
   resulting in a reported exit status of "1234567".
 * Fix for an error in system_script_forker that can cause an error loop
   if using stdout to string functionality.
 * Fix for an error that caused jobs with a "terminating" status to cause
   qstat to fail.

== Changes to 1.0.5 ==
 * Fix for an issue that would cause jobs with their location set to not
   run if they were submitted to a reservation.
 * Performance of setres improved significantly.
 * Fixed an issue where reservations and soft-down status on nodes may not
   be respected by job placement.
 * An enhanced_apkill script is now available to improve job cleanup and
   ensure that jobs that ignore signals are otherwise properly terminated.

== Changes to 1.0.4 ==
 * Fix for apid not being properly checked when apkill was being invoked.

== Changes to 1.0.3 ==
 * Fix for an issue for nodes coming out of disabled state and suddenly
   appearing in ALPS.
 * Interactive job support
 * Support for CAPMC memory management scripts added.  Cobalt now gracefully
   handles long startup times if these are being used.
 * Improved job cleanup using apkill.

== Changes to 1.0.2 ==
 * Singleton job filter (maxtotaljobs) now supported
 * Multiple alps_script_forkers running simultaneously is now supported. Forkers
   are chosen for execution based on which forker has the lowest number of jobs
   currently assigned to it.

== Changes to 1.0.0 ==
 * Initial support of Cray systems

== Changes to 0.99.39 ==
 * Session information added to job submission information.  This also includes
   submitting host information (Cobalt Ticket #4315)
 * A fix for a bug that can cause "phantom drains" to occur on lightly loaded
   cluster systems. (Cobalt Ticket #4319)

== Changes to 0.99.38 ==
 * Added numerous dependency changes.  Notably, jobs in dep_hold states will no
   longer accrue eligible wait time or score.  Previously, dep_hold jobs would
   accrue eligible wait times unlike admin_hold or user_hold states.
   (Cobalt Ticket #4308)
 * Two display fixes for cluster systems that allow multiple node-name prefixes
   to be handled correctly by the nodelist merging code  and to correctly
   compact names if leading 0's are used in location numbers like
   "foo001.machine". (Cobalt Ticket #4313 and #4277)
 * Corrected two race conditions in the BlueGene/Q system components where a
   location with offline hardware could be selected for draining on a
   scheduling pass.  Though Cobalt would never try to actually run on the
   offline block, It could end up draining resources needlessly, and impact
   machine usage.  The first could happen at any time, the second only occurred
   at the end of booting a block. (Cobalt Ticket #4311)

== Changes to 0.99.37 ==
 * Fixed issue where a reservation could hit it's activation check while it is being
   deactivated, causing those events should be reversed.  Released reservations can't
   activate. (Cobalt Ticket #4289)
 * COBALT_JOBSIZE and COBALT_PARTSIZE have been added to interactive jobs.
   (Cobalt Ticket #4294)
 * Job dependencies require the same project to gain score from eachother.
   (Cobalt Ticket #4293)
 * prologue_helper.py now handles "=" in arguments correctly. (Cobalt Ticket #4292)

== Changes to 0.99.36 ==
 * Fix for issues in get-bootable-blocks and with block booting.  Block names are now
   properly sorted.  A race condition between boot-block and update_block_state has
   been fixed.  The return statuses for boot-block have also been corrected for failed
   boots.  (Cobalt Ticket #4282 and #4280)

== Changes to 0.99.35 ==
 * Fixed an issue where a block being removed in the BGQ control system between block
   detection and the block being added for tracking could result in blocks being
   erroniously marked as idle. (Cobalt Ticket #4268)

== Changes to 0.99.34 ==
 * Fixed an issue where Cobalt could have a memory leak if boot-block.py was
   terminated early the boot tracking object may never be cleaned up.
 * Numerous enhancements and fixes to handle multiple overlaping queues in the
   system with respect to draining and backfilling. (Cobalt Ticket #663)
 * Fixed some typos in the documentation for qsub.
 * COBALT_RESID is now appropriately set in interactive mode jobs (Cobalt Ticket #703)

== Changes to 0.99.33 ==
 * Fixed issue where adding blocks while the system component is running can
   cause the system component's block update loop to fail. (Cobalt Ticket #695)
 * Preliminary fix for a racecondition that can cause subblock jobs to terminate
   early (Cobalt Ticket #690)
 * Added T-Minus to showres (Cobalt Ticket #692)
 * Fixed an error in qalter where attemtpts to change the users on a running job
   caused qalter to fail. (Cobalt Ticket #694)

== Changes to 0.99.32 ==
 * Cluster systems now drain for larger jobs, and will backfill into the
   draining hardware. (Cobalt Ticket #663)
 * Fixed a bug in the prologue_helper where a silent failure on sending the
   hostfile to nodes could occur silently (Cobalt Ticket #686)

== Changes to 0.99.31 == 
 * Interactive mode jobs (qsub -I) are now supported on the BlueGene/Q.
   (Cobalt Ticket #107)
 * Timestamps are now included in cobaltlog lines (Cobalt Ticket #344 and #669)
 * Multiple mode arguments will result in a qsub rejection (Cobalt Ticket #664)
 * Fixed an issue that was causing COBALT_STARTTIME and COBALT_ENDTIME to not be
    set for compute node jobs in cX jobs. (Cobalt Ticket #673)

== Changes to 0.99.30 ==
 * This is a fix release that corrects a bug that can be encountered when upgrading
   to 0.99.27 from prior versions where ion_kerneloptions in cqm is improperly set.
   (Cobalt Ticket #665)

== Changes to 0.99.29 ==

src/clients/nodelist.py
#!/usr/bin/env python
"""
Nodelist 

Usage: %prog
version: "%prog " + __revision__ + , Cobalt  + __version__

OPTIONS DEFINITIONS: None

'-d','--debug',dest='debug',help='turn on communication debugging',callback=cb_debug
'-b','--list_details',action='store_true', dest='list_details', help='list detalied information for specified nodes'
'--noheader',dest='noheader',action='store_true',help='disable display of header information'

"""
import logging
import sys
from Cobalt import client_utils
from Cobalt.client_utils import cb_debug

from Cobalt.arg_parser import ArgParse

__revision__ = 'TBD'
__version__ = 'TBD'

SYSMGR = client_utils.SYSMGR

def main():
    # setup logging for client. The clients should call this before doing anything else.
    client_utils.setup_logging(logging.INFO)

    # list of callback with its arguments
    callbacks = [
        # <cb function>     <cb args>
        [ cb_debug        , () ] ]

    # Get the version information
    opt_def =  __doc__.replace('__revision__',__revision__)
    opt_def =  opt_def.replace('__version__',__version__)

    parser = ArgParse(opt_def,callbacks)

    # Set required default values: None

    parser.parse_it() # parse the command line
    opt = parser.options
    args = parser.args

    #if not parser.no_args():
    #    client_utils.logger.error("No arguments needed")

    impl = client_utils.component_call(SYSMGR, False, 'get_implementation', ())

    # make sure we're on a cluster-system
    if impl not in ['cluster_system', 'alps_system']:
        client_utils.logger.error("nodelist is only supported on cluster systems.  Try partlist instead.")
        sys.exit(0)

    if impl == 'alps_system':
        if opt.list_details:
            # get list from arguments.  Currently assuing a comma separated,
            # hyphen-condensed nodelist
            client_utils.print_node_details(args)
        else:
            client_utils.print_node_list()
        return



    header, output = client_utils.cluster_display_node_info()
    if parser.options.noheader is not None:
        client_utils.printTabular(header + output, with_header_info=False)
    else:
        client_utils.printTabular(header + output)

if __name__ == '__main__':
    try:
        main()
    except SystemExit:
        raise
    except Exception, e:
        client_utils.logger.fatal("*** FATAL EXCEPTION: %s ***", e)
        sys.exit(1)
